### PR TITLE
feature/TASK-2_auth

### PR DESCRIPTION
**Pull request by JeriRov**

### CODE REVIEW

File: handler.ts [hunk 1]
```diff
@@ -0,0 +1,5 @@
+import { signUp } from './src/functions/signUp';
+import { test } from './src/functions/test';
+import { signIn } from './src/functions/signIn';
+
+export { signUp, test, signIn };
```
Comment (milkanuch) [inline handler.ts]: you can also write:
`export { signUp } from '...'`

File: src/types/MiddyCustom.ts [hunk 2]
```diff
@@ -0,0 +1,4 @@
+import { APIGatewayProxyEvent } from 'aws-lambda';
+import { UserWithoutPassword } from './User';
+
+export type MiddyEvent<T = APIGatewayProxyEvent['body']> = APIGatewayProxyEvent & { body: T, auth?: UserWithoutPassword | null };
```
Comment (milkanuch) [inline src/types/MiddyCustom.ts]: Please, write body and auth in separate lines:

```
export type MiddyEvent<T = APIGatewayProxyEvent['body']> = 
APIGatewayProxyEvent & {
 body: T, 
 auth?: UserWithoutPassword | null 
};```

File: src/middlewares/authenticate.ts [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+import { APIGatewayProxyEvent, APIGatewayProxyResult } from 'aws-lambda';
+import createHttpError from 'http-errors';
+import middy, { MiddlewareObj } from '@middy/core';
+import { UserWithoutPassword } from '../types/User';
+import { MiddyEvent } from '../types/MiddyCustom';
+
+const JWT_SECRET = process.env.JWT_SECRET as Secret;
+
+export const authenticate = (): MiddlewareObj<APIGatewayProxyEvent, APIGatewayProxyResult> => {
+  const before: middy.MiddlewareFn<MiddyEvent, APIGatewayProxyResult> = (
```
Comment (milkanuch) [inline src/middlewares/authenticate.ts]: I think we can take out this inner function, and avoid creating it on every authenticate call

File: src/helpers/tableNames.ts [hunk 4]
```diff
@@ -0,0 +1,3 @@
+export const TableNames = {
+  USER: process.env.USER_TABLE_NAME || '',
```
Comment (milkanuch) [inline src/helpers/tableNames.ts]: if you trying to avoid mutations, use Object.freeze

File: src/functions/signUp.ts [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+  const { email, password } = event.body;
+  if (!email || !password) {
+    throw new createHttpError.Conflict('Email and password are required!');
+  }
+  if (!validateEmail(email)) {
+    throw new createHttpError.Conflict('Invalid email');
+  }
+  if (!validatePassword(password)) {
+    throw new createHttpError.Conflict('The password must contain a minimum of eight characters, at least one letter and one number');
+  }
```
Comment (milkanuch) [inline src/functions/signUp.ts]: Make this function reusable with arg avoidPassword

**Review by milkanuch — APPROVED** [review]

File: handler.ts (hunk 1 above)
Comment (JeriRov) [reply handler.ts]: Thanks, already changed!

File: src/types/MiddyCustom.ts (hunk 2 above)
Comment (JeriRov) [reply src/types/MiddyCustom.ts]: Thanks, I'll keep that in mind

File: src/middlewares/authenticate.ts (hunk 3 above)
Comment (JeriRov) [reply src/middlewares/authenticate.ts]: Sorry, I was doing this according to the [middy](https://middy.js.org/docs/writing-middlewares/with-typescript) documentation, so I don't think I can take out this inner function.

File: src/helpers/tableNames.ts (hunk 4 above)
Comment (JeriRov) [reply src/helpers/tableNames.ts]: Somehow?
```typescript
export const TableNames = Object.freeze({
  USER: process.env.USER_TABLE_NAME ?? '',
});
```

File: src/functions/signUp.ts (hunk 5 above)
Comment (JeriRov) [reply src/functions/signUp.ts]: Ok, did it. 👀

**Review by milkanuch — COMMENTED** [review]
LGTM!

**Pull request merged**